### PR TITLE
Meta reshape

### PR DESCRIPTION
**Pull request by aburrell**
Reshape of Meta class, introduces new MetaLabels class

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality
      to not work as expected)
- This change requires a documentation update

**This PR will be closed after review/updates** <- a LIE!

@rstoneback I have a problem where metadata that should be set isn't.  Not sure how to fix this!  I tracked it down to lines 521-543 of `_instrument.py` and need you to take a look at it.

Example of a thing that fails:  `python3 -m pytest test_meta.py::TestBasics::test_inst_data_assign_meta`

## Checklist:
- [x] Make sure you are merging into the develop (not main) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x]  Any dependent changes have been merged and published in downstream modules
- [x]  Add a note to CHANGELOG.md, summarizing the changes

### Downstream modules:
- [x] pysatSpaceWeather: https://github.com/pysat/pysatSpaceWeather/pull/28
- [x] pysatMissions: pysat/pysatMissions#41
- [x] pysatNASA: https://github.com/pysat/pysatNASA/pull/25
- [x] pysatIncubator: pysat/pysatIncubator#6
- [x] pysatModels: https://github.com/pysat/pysatModels/pull/69
- [x] pysatMadrigal: https://github.com/pysat/pysatMadrigal/pull/28
- [x] pysatCDAAC: pysat/pysatCDAAC#11
- [x] pysatSeasons: @rstoneback 
- [x] pysatCDF: Created new develop-3 branch for changes needed for the pysat 3.0.0 release.

### CODE REVIEW

File: pysat/_meta.py [hunk 1]
```diff
@@ -477,19 +249,6 @@ def __str__(self, long_str=True):
...
 
-    def _insert_default_values(self, input_name):
-
-        default_str = ''
-        default_nan = np.NaN
-        labels = [self.units_label, self.name_label, self.notes_label,
-                  self.desc_label, self.plot_label, self.axis_label,
-                  self.scale_label, self.min_label, self.max_label,
-                  self.fill_label]
-        defaults = [default_str, input_name, default_str, default_str,
```
Comment (rstoneback) [inline pysat/_meta.py:488]: I believe this is the issue. The default for some of the metadata parameters is set to `input_name`. When I traced through the current code I don't see this functionality replicated.

File: pysat/_instrument.py [hunk 2]
```diff
@@ -175,11 +175,13 @@ def __init__(self, platform=None, name=None, tag=None, inst_id=None,
...
                  temporary_file_list=False, strict_time_flag=True,
                  ignore_empty_files=False,
-                 units_label='units', name_label='long_name',
-                 notes_label='notes', desc_label='desc',
-                 plot_label='label', axis_label='axis', scale_label='scale',
-                 min_label='value_min', max_label='value_max',
-                 fill_label='fill', *arg, **kwargs):
+                 labels={'units': ('units', str), 'name': ('long_name', str),
+                         'notes': ('notes', str), 'desc': ('desc', str),
+                         'plot': ('plot_label', str), 'axis': ('axis', str),
```
Comment (rstoneback) [inline pysat/_instrument.py:180]: I don't think 'label' and 'plot_label' is consistent everywhere for `'plot`. I'm getting 'label' and 'plot_label' for your latest tests.

```
pysat/tests/test_meta.py::TestBasics::test_inst_data_assign_meta[labels2-vals2]

meta 
 units                                     
long_name       test_inst_data_assign_meta
notes                                     
desc                                      
plot_label      test_inst_data_assign_meta
axis            test_inst_data_assign_meta
scale                               linear
value_min                              NaN
value_max                              NaN
fill                                  -999
custom                                 NaN
Bin_Location                           NaN
label                                  NaN
children                              None
Name: test_inst_data_assign_meta, dtype: object

```

Comment (rstoneback) [reply pysat/_instrument.py:180]: I also had to change a few lines in test_meta.py:

```
        self.default_name = ['long_name', 'axis', 'plot_label']

```
under `TestBasicsImmutable` and `TestBasics`.

File: pysat/_meta.py [hunk 3]
```diff
@@ -1375,3 +1243,261 @@ def from_csv(cls, name=None, col_names=None, sep=None, **kwargs):
...
+        String used to label units in storage. (default='units'.
+    name_label : str
+        String used to label long_name in storage. (default='long_name'.
+    notes_label : str
+       String used to label 'notes' in storage. (default='notes'
+    desc_label : str
+       String used to label variable descriptions in storage.
+       (default='desc'
+    plot_label : str
+       String used to label variables in plots. (default='label'
```
Comment (rstoneback) [inline pysat/_meta.py:1297]: Check default of __init__ call

**Review by rstoneback — COMMENTED** [review]
I made some progress on the defaults. I'm able to get past the 
`
python -m pytest -vs pysat/tests/test_meta.py::TestBasics::test_inst_data_assign_meta
`
with the included suggestions.

I'm seeing both 'label' and 'plot_label' in the meta data object in that test. I haven't sorted out why there are two variables for the plot_label. I think there are a couple areas where the docstring listed default isn't the same as the actual default. Maybe there is a leftover code area I didn't see that is still using 'label'.

I don't get past all the tests. If that isn't expected then I broke something.

File: pysat/_instrument.py (hunk 2 above)
Comment (aburrell) [reply pysat/_instrument.py:180]: It looks like `plot_label` is assigned by default, but `label` is being specifically asked for by that test.  So it's testing the setting of a custom MetaLabel in this instance.

File: pysat/_meta.py (hunk 3 above)
Comment (aburrell) [inline pysat/_meta.py:1297]:
```suggestion
       String used to label variables in plots. (default='plot'
```

File: pysat/_meta.py [hunk 4]
```diff
@@ -1375,3 +1243,261 @@ def from_csv(cls, name=None, col_names=None, sep=None, **kwargs):
...
+        String used to label units in storage. (default='units')
+    name : str
+        String used to label name in storage. (default='long_name')
+    notes : str
+        String used to label 'notes' in storage. (default='notes')
+    desc : str
+        String used to label variable descriptions in storage.
+        (default='desc')
+    plot : str
+        String used to label variables in plots. (default='label')
```
Comment (aburrell) [inline pysat/_meta.py:1263]:
```suggestion
        String used to label variables in plots. (default='plot')
```

File: pysat/_meta.py [hunk 5]
```diff
@@ -764,163 +534,387 @@ def match_name(func, name, names):
...
+            raise NotImplementedError("".join(["No way to handle MetaData key ",
+                                               "{}; ".format(key.__repr__()),
+                                               "expected tuple, list, or str"]))
 
-    def _label_setter(self, new_label, current_label, attr_label,
-                      default=np.nan, use_names_default=False):
-        """Generalized setter of default meta attributes
+    # QUESTION: DOES THIS NEED TO CHANGE???
+    def __contains__(self, other):
+        """case insensitive check for variable name
```
Comment (aburrell) [inline pysat/_meta.py:549]: @rstoneback I am not sure what this routine is supposed to do.  Is it behaving correctly?

File: pysat/_meta.py [hunk 6]
```diff
@@ -1375,3 +1243,261 @@ def from_csv(cls, name=None, col_names=None, sep=None, **kwargs):
...
+        # Test the input parameter
+        if attr_name not in self.label_type.keys():
+            raise ValueError('uknown label attribute {:}'.format(attr_name))
+
+        # Assign the default value
+        if attr_name == 'scale':
+            default_val = 'linear'
+        else:
+            default_val = self.default_values_from_type(
+                self.label_type[attr_name])
```
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/_meta.py:1499]: I changed this elsewhere.

**Review by aburrell — COMMENTED** [review]
Ok @rstoneback, thanks to your insight I was able to make the suggested change at (what I think) is the appropriate level in the classes.  I also went through and fixed other stray bugs.  Now, all unit tests are passing but 11 in `test_meta`, which are related to a single problem.  When you update the labels (e.g., meta.labels.units = 'Units'), the change isn't propagated through to the pandas Series.  I'm not sure how this was accomplished before.

```
=========================== short test summary info ============================
FAILED test_meta.py::TestBasics::test_change_case_of_meta_labels - AttributeE...
FAILED test_meta.py::TestBasics::test_change_Units_and_Name_case_w_ho - Attri...
FAILED test_meta.py::TestBasics::test_change_Units_and_Name_case_w_ho_wrong_case
FAILED test_meta.py::TestBasicsImmutable::test_assign_capitalized_labels - At...
FAILED test_meta.py::TestBasicsImmutable::test_assign_Units_no_units - Attrib...
FAILED test_meta.py::TestBasicsImmutable::test_get_Units_wrong_case - Attribu...
FAILED test_meta.py::TestBasicsImmutable::test_set_Units_wrong_case - Attribu...
FAILED test_meta.py::TestBasicsImmutable::test_repeated_set_Units_wrong_case
FAILED test_meta.py::TestBasicsImmutable::test_change_case_of_meta_labels - A...
FAILED test_meta.py::TestBasicsImmutable::test_change_Units_and_Name_case_w_ho
FAILED test_meta.py::TestBasicsImmutable::test_change_Units_and_Name_case_w_ho_wrong_case
```


I also tagged you in a comment about a method whose behaviour I don't understand.

File: pysat/_meta.py [hunk 7]
```diff
@@ -764,163 +534,387 @@ def match_name(func, name, names):
...
-        self._label_setter(new_label, self._min_label, 'min_label', np.NaN)
+            # now iterate over each of the keys in the first one
+            # don't need to iterate over second one, if all of the first
+            # in the second we are good. No more or less items in second from
+            # check earlier.
+            for key in keys1:
+                if key not in keys2:
+                    return False
 
-    @max_label.setter
```
Comment (rstoneback) [inline pysat/_meta.py:915]: The labels used to actively update the underlying pandas data object whenever a label was changed by calling the generalized self._label_setter.

**Review by rstoneback — COMMENTED** [review]
Some of the failures you listed come up since `TestBasicsImmutable` is missing a
```
        self.meta_labels = {'units': ('Units', str),
                            'name': ('Long_Name', str)}
```
in setup. Adding this here since I can't comment on unchanged lines.

I'm also getting a
`test_meta_csv_load_w_errors[name-None-Must provide an instrument] - TypeError: parser_f() got an unexpected keyword argument 'name'` error. May be a pandas version difference.

File: pysat/_meta.py (hunk 5 above)
Comment (rstoneback) [reply pysat/_meta.py:549]: > Ok @rstoneback, thanks to your insight I was able to make the suggested change at (what I think) is the appropriate level in the classes. I also went through and fixed other stray bugs. Now, all unit tests are passing but 11 in `test_meta`, which are related to a single problem. When you update the labels (e.g., meta.labels.units = 'Units'), the change isn't propagated through to the pandas Series. I'm not sure how this was accomplished before.
> 


Check out the old meta method self._label_setter. I pointed to the general area in the code highlight a few comments down.

File: pysat/_meta.py [hunk 8]
```diff
@@ -764,163 +534,387 @@ def match_name(func, name, names):
...
-        Examples
-        --------
-        :
-
-            @name_label.setter
-            def name_label(self, new_label):
-                self._label_setter(new_label, self._name_label,
-                                    use_names_default=True)
+        other : Meta
+            Meta object from which the default labels are obtained
```
Comment (rstoneback) [inline pysat/_meta.py:554]:
```suggestion
        other : str
            Variable name to check if present within the Meta object.
```

File: pysat/_meta.py (hunk 5 above)
Comment (rstoneback) [reply pysat/_meta.py:549]: > @rstoneback I am not sure what this routine is supposed to do. Is it behaving correctly?

The method lets you know if the string provided in other is a variable in the meta object. The string comparison is case insensitive. Other than the docstring looks ok to me.

```
In [8]: In [2]: test = pysat.Instrument('pysat', 'testing')                                                                                                                                                                 
   ...:  
   ...: In [3]: test.load(2009, 1)                                                                                                                                                                                          
   ...:                                                                                                                                                                                                                     

In [9]: 'uts' in test.meta                                                                                                                                                                                                  
contains!
Out[9]: True

In [10]: 'UTS' in test.meta                                                                                                                                                                                                 
contains!
Out[10]: True

In [11]: test.variables                                                                                                                                                                                                     
Out[11]: 
Index(['uts', 'mlt', 'slt', 'longitude', 'latitude', 'altitude', 'orbit_num',
       'dummy1', 'dummy2', 'dummy3', 'dummy4', 'string_dummy', 'unicode_dummy',
       'int8_dummy', 'int16_dummy', 'int32_dummy', 'int64_dummy'],
      dtype='object')
```

**Comment by rstoneback** [comment]
I only focused on your specific questions due to time constraints.

**Comment by aburrell** [comment]
Ok @rstoneback more improvement has been made.  However, it's still broke.  I have print statements that say "HI!", "FUN", and "TEST" in areas of interest.  There are two problems:

1) Because labels aren't properties anymore the mutable behaviour is different.  Do we want them to be immutable if the mutable flag is false?  Or do we want to add labels as an exception?
2) Recursive updates to the child Series aren't working.  I tried a few things I thought would work, but couldn't nail it.

Thoughts and ideas?

**Comment by rstoneback** [comment]
Thanks @aburrell.

> Ok @rstoneback more improvement has been made. However, it's still broke. I have print statements that say "HI!", "FUN", and "TEST" in areas of interest. There are two problems:
> 
> 1. Because labels aren't properties anymore the mutable behaviour is different.  Do we want them to be immutable if the mutable flag is false?  Or do we want to add labels as an exception?

I hope you don't mind a little background. The mutable flag was added to keep people from adding random things to meta since those random things will end up getting thrown away upon the next load. If changing the labels is stored then they should be an exception. Best answer is it probably depends upon 2.

> 2. Recursive updates to the child Series aren't working.  I tried a few things I thought would work, but couldn't nail it.
> 
> Thoughts and ideas?

I'll give it a look and some time.

I'm tied up with a site review through the end of the week. I'm likely only going to be good for small stuff the next couple days.

File: pysat/_instrument.py [hunk 9]
```diff
@@ -265,27 +255,11 @@ def __init__(self, platform=None, name=None, tag=None, inst_id=None,
...
-                               axis_label=self.axis_label,
-                               scale_label=self.scale_label,
-                               min_label=self.min_label,
-                               max_label=self.max_label,
-                               fill_label=self.fill_label)
+        # Create Meta instance with appropriate labels.  Meta class methos will
+        # use Instrument definition of MetaLabels over the Metadata declaration
+        self.meta_labels = labels
+        self.meta = pysat.Meta(labels=self.meta_labels)
+        self.labels = pysat.MetaLabels(metadata=self.meta, **labels)
```
Comment (rstoneback) [inline pysat/_instrument.py:262]: Initial impression. Shouldn't we pass in MetaLabels to Meta to define the labels, rather than the dict?

File: pysat/_instrument.py [hunk 10]
```diff
@@ -2489,16 +2437,16 @@ def to_netcdf4(self, fname=None, base_instrument=None, epoch_name='Epoch',
...
             # didn't find a translation table, using the strings
             # attached to the supplied pysat.Instrument object
-            export_name_labels = [self.name_label]
-            export_units_labels = [self.units_label]
-            export_desc_labels = [self.desc_label]
-            export_notes_labels = [self.notes_label]
+            export_name_labels = [self.meta_labels['name'][0]]
+            export_units_labels = [self.meta_labels['units'][0]]
+            export_desc_labels = [self.meta_labels['desc'][0]]
+            export_notes_labels = [self.meta_labels['notes'][0]]
```
Comment (rstoneback) [inline pysat/_instrument.py:2443]: Does this work?

```suggestion
            export_name_labels = [self.labels.name]
            export_units_labels = [self.labels.units]
            export_desc_labels = [self.labels.desc]
            export_notes_labels = [self.labels.notes]
```

File: pysat/tests/test_custom.py [hunk 11]
```diff
@@ -251,8 +251,8 @@ def custom1(inst):
 
         self.testInst.custom.attach(custom1, 'add')
         self.testInst.load(2009, 1)
-        assert self.testInst.meta['doubleMLT'].units == 'hours1'
-        assert self.testInst.meta['doubleMLT'].long_name == 'doubleMLT'
+        assert self.testInst.meta[('doubleMLT', 'units')] == 'hours1'
+        assert self.testInst.meta[('doubleMLT', 'long_name')] == 'doubleMLT'
```
Comment (rstoneback) [inline pysat/tests/test_custom.py:255]:
```suggestion
        assert self.testInst.meta['doubleMLT', 'units'] == 'hours1'
        assert self.testInst.meta['doubleMLT', 'long_name'] == 'doubleMLT'
```

File: pysat/_meta.py [hunk 12]
```diff
@@ -111,93 +72,89 @@ class Meta(object):
...
-        self._max_label = max_label
-        self._fill_label = fill_label
-
-        # by default metadata with a value of nan will not be exported
-        # unless the name is in the _export_nan list. Initialize the list
-        # with the fill label, since it is reasonable to assume that a fill
-        # value of nan would be intended to be exported
-        self._export_nan = [fill_label] + export_nan
+        # Set the NaN export list
+        self._export_nan = export_nan
```
Comment (rstoneback) [inline pysat/_meta.py:152]: Why did you remove fill_val as a default for NaN support?

**Review by rstoneback — COMMENTED** [review]
Thanks @aburrell! I've made a partial pass through. I will be back to spend more time on the changes. Cheers.

File: pysat/_instrument.py (hunk 9 above)
Comment (aburrell) [reply pysat/_instrument.py:262]: Nope.  It requires the dict.  Not sure if we really need the labels definition here, but I'm trying not to break everything :P

File: pysat/_instrument.py (hunk 10 above)
Comment (aburrell) [reply pysat/_instrument.py:2443]: Probably?  This would be cleaner, for sure.

Comment (aburrell) [reply pysat/_instrument.py:2443]: Or perhaps it would be better to do `self.meta.labels.name`, et cetera.

File: pysat/_meta.py (hunk 12 above)
Comment (aburrell) [reply pysat/_meta.py:152]: Originally I did not see what this did, so I had removed it and added it back in.  As is, any float value should all have NaN support.

File: pysat/tests/test_custom.py (hunk 11 above)
Comment (aburrell) [reply pysat/tests/test_custom.py:255]: Was it not working for you, or is this a style thing?

File: pysat/_instrument.py (hunk 10 above)
Comment (aburrell) [reply pysat/_instrument.py:2443]: Yeah, that's better style.

**Comment by rstoneback** [comment]
Starting in on this. I'm getting that changing, say `meta.labels.name = 'Long_Name'` works, but it only works once. The second time an update is attempted I get a 
```
In [21]: meta.labels.units = 'help'                                                                                                                                                                                         
FUN! units
<super: <class 'MetaLabels'>, <MetaLabels object>>
FUN! metadata
<super: <class 'MetaLabels'>, <MetaLabels object>>
TEST units
HI! new2 units help : <class 'pysat._meta.Meta'> MetaLabels:
-----------
units->help        name->long_Name    notes->notes       
desc->desc         plot->plot         axis->axis         
scale->scale       min_val->value_min max_val->value_max 
fill_val->fill     
 Index(['units', 'long_Name', 'notes', 'desc', 'plot', 'axis', 'scale',
       'value_min', 'value_max', 'fill'],
      dtype='object') [<class 'pandas.core.series.Series'>]
TEST _units

In [24]: meta.labels.units = 'Units'                                                                                                                                                                                        
FUN! units
<super: <class 'MetaLabels'>, <MetaLabels object>>
TEST data
TEST mutable
TEST _data
TEST mutable
FUN! metadata
<super: <class 'MetaLabels'>, <MetaLabels object>>
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-24-8e525d030d72> in <module>
----> 1 meta.labels.units = 'Units'

~/Code/pysat/pysat/_meta.py in __setattr__(self, name, value)
   1438         if old_value is not None and name not in ['label_type', 'meta']:
   1439             if hasattr(self, 'meta') and hasattr(self.meta, 'data'):
-> 1440                 self.meta._label_setter(value, getattr(self, name), name,
   1441                                         self.label_type[name],
   1442                                         use_names_default=True)

~/Code/pysat/pysat/_meta.py in _label_setter(self, new_label, current_label, attr_label, default_type, use_names_default)
    782             for key in self.keys_nD():
    783                 # Recursion isn't working, need to update children
--> 784                 self.ho_data[key].labels.metadata = self.ho_data[key]
    785                 setattr(self.ho_data[key], attr_label, new_label)
    786 

~/Code/pysat/pysat/_meta.py in __setattr__(self, name, value)
   1439             if hasattr(self, 'meta') and hasattr(self.meta, 'data'):
   1440                 self.meta._label_setter(value, getattr(self, name), name,
-> 1441                                         self.label_type[name],
   1442                                         use_names_default=True)
   1443 

KeyError: 'metadata'
> /Users/russellstoneback/Code/pysat/pysat/_meta.py(1441)__setattr__()
   1439             if hasattr(self, 'meta') and hasattr(self.meta, 'data'):
   1440                 self.meta._label_setter(value, getattr(self, name), name,
-> 1441                                         self.label_type[name],
   1442                                         use_names_default=True)
   1443 

```
Things are taking a different path the second time round.

**Comment by rstoneback** [comment]
I think I sorted out where the problem is.

Incidentally, I've wanted to use super the way you do but the docs I've read indicated it wouldn't work. Good to know. Thanks.

**Comment by rstoneback** [comment]
I also 'fixed' `accept_default_labels` but that broke tests. I've commented out those changes while I think about what I've done.

**Comment by rstoneback** [comment]
> I think I sorted out where the problem is.
> 
> Incidentally, I've wanted to use super the way you do but the docs I've read indicated it wouldn't work. Good to know. Thanks.

It wasn't doing what I expected.

**Comment by rstoneback** [comment]
I think that should cover it. Tests passed last round except for flake8.

**Comment by aburrell** [comment]
Thanks @rstoneback I will take a look at this later (today?) and address the outstanding review points.

**Comment by rstoneback** [comment]
> Thanks @rstoneback I will take a look at this later (today?) and address the outstanding review points.

You are most welcome @aburrell. I certainly can't say anything about timetables.

**Comment by rstoneback** [comment]
A test error came up from one of my pushes so I thought it reasonable to fix it.

**Review by rstoneback — DISMISSED** [review]
Thanks @aburrell!

**Comment by rstoneback** [comment]
I see that :P I have a meeting coming up here in a couple minutes. I'll take a pass through after.

File: pysat/_instrument.py [hunk 13]
```diff
@@ -2632,18 +2585,24 @@ def load(self, yr=None, doy=None, end_yr=None, end_doy=None, date=None,
...
+                    if case_var not in self.meta.keys() \
+                       and case_var not in self.meta.keys_nD():
+                        self.meta[case_var] = {self.labels.name: var}
+
         # check if load routine actually returns meta
         if self.meta.data.empty:
-            self.meta[self.variables] = {self.labels.name: self.variables,
-                                         self.labels.units:
-                                         [''] * len(self.variables)}
+            self.meta[self.variables] = {self.labels.name: self.variables}
```
Comment (rstoneback) [inline pysat/_instrument.py:2605]: The same could be said for this statement. I think I originally added it before unit tests. 

I'm not opposed to a soft-floor for standards to help users out but given how standards have steadily increased for pysat we may want to modify the approach.

Comment (rstoneback) [reply pysat/_instrument.py:2605]: This was posted in a weird order. The `The same could be said for this statement` refers to the check being a unit test.

File: pysat/_instrument.py [hunk 14]
```diff
@@ -45,21 +45,24 @@ class Instrument(object):
...
+    strict_time_flag : boolean
         If true, pysat will check data to ensure times are unique and
-        monotonically increasing.
-    multi_file_day : boolean, optional
+        monotonically increasing. (default=True)
+    multi_file_day : boolean or NoneType
         Set to True if Instrument data files for a day are spread across
         multiple files and data for day n could be found in a file
-        with a timestamp of day n-1 or n+1.
+        with a timestamp of day n-1 or n+1.  (default=None)
```
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/_instrument.py:65]: So, here's a question.  Why is the default not just False?  The way you describe it seems like we just make the code more complicated and costly to run.  However, I finally figured out why you use None for an empty list so can be convinced otherwise.

File: pysat/_instrument.py [hunk 15]
```diff
@@ -2632,18 +2585,24 @@ def load(self, yr=None, doy=None, end_yr=None, end_doy=None, date=None,
                 if (self.index[-1] == last_pad) & (not want_last_pad):
                     self.data = self[:-1]
 
-        # if self.pad is False, load single day
+        # If self.pad is False, load single day
         else:
             self.data, meta = self._load_data(date=self.date, fid=self._fid,
                                               inc=self.load_step)
             if not self.empty:
+                meta.accept_default_labels(self.meta)
```
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/_instrument.py:2593]: No, but also wasn't doing that in all cases.  I think it is actually cleaner to pass labels to load, and I intend to add that as a recommended optional kwarg in the docs, but that's something to discuss in a meeting.

File: pysat/_instrument.py [hunk 16]
```diff
@@ -1513,61 +1516,8 @@ def bounds(self):
 
     @bounds.setter
     def bounds(self, value=None):
-        """Sets the self.bounds property.
+        # Set the bounds property.  See property docstring for details
 
-        Parameters
```
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/_instrument.py:1568]: I would have, but I learned something new going over the CI doc tests :)

File: pysat/_instrument.py (hunk 13 above)
Comment (aburrell) [reply pysat/_instrument.py:2605]: This statement is useful, but your comments are a bit muddled and the clarification not as clear as I would like to be able to respond.  Perhaps we should have a chat about your comments?

File: pysat/_instrument.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_instrument.py:65]: It started from the first python conference I was at that suggested using None in the method signatures for protection against unintended consequences. Particularly for items that can be modified in memory. Comes with the downside of kinda killing the transparency of the default value feature.

For this particular case, the None is used to distinguish between the user leaving things as default or specifying a False value. If left as default, then multi_file_day is set to False unless the instrument method provides its own value. A user supplied value overrides those.

Primarily though I use a None whenever a default value is gong to be an empty list or string.

File: pysat/_instrument.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_instrument.py:1568]: Good to know. I find it a challenge keeping it all in my head at once. I do periodically review the standards but unfortunately details keep leaking out.

File: pysat/_instrument.py (hunk 13 above)
Comment (rstoneback) [reply pysat/_instrument.py:2605]: Sure.

File: pysat/_instrument.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_instrument.py:2593]: Integers in meta revealed a hole in `_load_data` that I will check out.

File: pysat/instruments/pysat_testing.py [hunk 17]
```diff
@@ -250,73 +249,49 @@ def load(fnames, tag=None, inst_id=None, sim_multi_file_right=False,
 
     data.index = index
     data.index.name = 'Epoch'
-    return data, meta.copy()
+
```
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/instruments/pysat_testing.py:252]: Addressed in Issue #621

File: pysat/_instrument.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_instrument.py:2593]: Tests still pass locally for me without this statement. I also ran through the integer test manually and didn't run into an issue.

I printed out the metadata just after each of the current `accept_default_labels` calls but I don't see a difference.
```
Normal accept labels
pysat Meta object
-----------------
Tracking 12 metadata values
Metadata for 8 standard variables
Metadata for 0 ND variables

MetaLabels:
-----------
units->Units       name->Long_Name    notes->notes       
desc->desc         plot->plot         axis->axis         
scale->scale       min_val->value_min max_val->value_max 
fill_val->fill     

Standard Metadata variables:
uts       Epoch     mlt       
             ...              
longitude latitude  altitude  


second accept labels
pysat Meta object
-----------------
Tracking 12 metadata values
Metadata for 8 standard variables
Metadata for 0 ND variables

MetaLabels:
-----------
units->Units       name->Long_Name    notes->notes       
desc->desc         plot->plot         axis->axis         
scale->scale       min_val->value_min max_val->value_max 
fill_val->fill     

Standard Metadata variables:
uts       Epoch     mlt       
             ...              
longitude latitude  altitude  


```

[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/_instrument.py:2593]: You need to print out the `repr` to see if `fill_val` is an int or not.  Of course, it's also possible I made a mistake.

[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_instrument.py:2593]: Will do. Thanks for the guidance.

[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_instrument.py:2593]: Here is what I'm getting both with and without this second `accept_default_labels`
```
In [1]: default_val = {'notes': '', 'units': '', 'desc': '', 
   ...:                             'scale': 'linear'}                                                                                                                                                                      

In [2]: default_name = ['long_name', 'axis', 'plot']                                                                                                                                                                        

In [3]: default_nan = ['fill', 'value_min', 'value_max']                                                                                                                                                                    

In [4]: import numpy as np                                                                                                                                                                                                  

In [5]: import pysat                                                                                                                                                                                                        

In [7]: meta_labels = {'units': ('Units', str), 
   ...:                             'name': ('Long_Name', str)}                                                                                                                                                             

In [8]:          meta_labels['fill_val'] = ("fill", int) 
   ...:          testInst = pysat.Instrument('pysat', 'testing', 
   ...:                                           clean_level='clean', 
   ...:                                           labels=meta_labels) 
   ...:          testInst.load(2009, 1) 
   ...:          meta = testInst.meta 
   ...:          dval = 'int32_dummy' 
   ...:  
   ...:           # Update the testing data 
   ...:          default_val['fill'] = -1 
   ...:          default_nan.pop(default_nan.index('fill')) 

In [9]:         for lkey in default_name: 
   ...:             assert meta[dval, lkey] == dval 
   ...:  
   ...:         for lkey in default_nan: 
   ...:             assert np.isnan(meta[dval, lkey]) 
   ...:  
   ...:         for lkey in default_val.keys(): 
   ...:             assert meta[dval, lkey] == default_val[lkey] 
   ...:  
   ...:         assert 'children' not in meta.data.columns 
   ...:         assert dval not in meta.keys_nD()                                                                                                                                                                           

In [11]: testInst.meta.data['fill']                                                                                                                                                                                         
Out[11]: 
uts              NaN
Epoch            NaN
mlt              NaN
slt              NaN
orbit_num        NaN
longitude        NaN
latitude         NaN
altitude         NaN
dummy1            -1
dummy2            -1
dummy3            -1
dummy4            -1
string_dummy      -1
unicode_dummy     -1
int8_dummy        -1
int16_dummy       -1
int32_dummy       -1
int64_dummy       -1
Name: fill, dtype: object
```

All of the integer data is getting -1 for the fill value. Floats are getting NaN (I haven't sorted it how it knows yet, traced a couple functions) but the strings are also getting -1. If the floats retain a NaN for fill despite a -1 default, shouldn't strings get a ''? 

I checked just a straight load (develop-3 branch) and got,
```
In [5]: test.meta.data['fill']                                                                                                                                                                                              
Out[5]: 
uts              NaN
Epoch            NaN
mlt              NaN
slt              NaN
orbit_num        NaN
longitude        NaN
latitude         NaN
altitude         NaN
dummy1           NaN
dummy2           NaN
dummy3           NaN
dummy4           NaN
string_dummy     NaN
unicode_dummy    NaN
int8_dummy       NaN
int16_dummy      NaN
int32_dummy      NaN
int64_dummy      NaN
Name: fill, dtype: object
```
This point about strings is a pre-existing condition and probably out of scope. If you have a solution though....

On point, I haven't found a difference with the second `accept_default_labels` and should be removed. Unit tests pass both ways.

[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/_instrument.py:2593]: Ah, this is because meta default fill values don't take the variable type into account, only the label type into account.  That is definitely out of scope here and would change the way meta works.

File: pysat/_instrument.py [hunk 18]
```diff
@@ -2682,18 +2585,24 @@ def load(self, yr=None, doy=None, end_yr=None, end_doy=None, date=None,
                 if (self.index[-1] == last_pad) & (not want_last_pad):
                     self.data = self[:-1]
 
-        # if self.pad is False, load single day
+        # If self.pad is False, load single day
         else:
             self.data, meta = self._load_data(date=self.date, fid=self._fid,
                                               inc=self.load_step)
             if not self.empty:
+                meta.accept_default_labels(self.meta)
```
Comment (aburrell) [inline pysat/_instrument.py:2593]:
```suggestion
```

File: pysat/_instrument.py [hunk 19]
```diff
@@ -1816,9 +1716,9 @@ def date(self):
         return self._date
 
     @date.setter
-    def date(self, new):
-        """Date for loaded data."""
-        self._date = self._filter_datetime_input(new)
+    def date(self, new_date):
+        # Set the date property, see property docstriing for details
```
Comment (aburrell) [inline pysat/_instrument.py:1720]:
```suggestion
        # Set the date property, see property docstring for details
```

File: pysat/_instrument.py [hunk 20]
```diff
@@ -2632,18 +2585,24 @@ def load(self, yr=None, doy=None, end_yr=None, end_doy=None, date=None,
...
             if not self.empty:
+                meta.accept_default_labels(self.meta)
                 self.meta = meta
 
+                # If only some metadata included, define the remaining variables
+                for var in self.variables:
+                    case_var = meta.var_case_name(var)
+                    if case_var not in self.meta.keys() \
+                       and case_var not in self.meta.keys_nD():
+                        self.meta[case_var] = {self.labels.name: var}
```
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/_instrument.py:2620]: @rstoneback what did we decide to do about this?

**Comment by jklenzing** [comment]
Found un-pushed instrument style update branch in pysatCDAAC, so I went ahead and added the meta changes there.

File: pysat/_instrument.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_instrument.py:2620]: I think we landed on putting up a warning that some meta was missing, then add it in for them.

**Comment by aburrell** [comment]
@rstoneback it's fixed now, ready for approval or more improvements.

**Review by rstoneback — APPROVED** [review]
Thanks @aburrell! While there is always more to do this is a good pull.
